core/trino-main/src/main/java/io/trino/execution/SqlTaskExecution.java

```diff
@@ -602,30 +563,71 @@ public Driver createDriver(DriverContext driverContext, @Nullable ScheduledSplit
         {
             Driver driver = driverFactory.createDriver(driverContext);
 
-            // record driver so other threads add unpartitioned sources can see the driver
-            // NOTE: this MUST be done before reading unpartitionedSources, so we see a consistent view of the unpartitioned sources
-            drivers.add(new WeakReference<>(driver));
-
             if (partitionedSplit != null) {
                 // TableScanOperator requires partitioned split to be added before the first call to process
                 driver.updateSplitAssignment(new SplitAssignment(partitionedSplit.getPlanNodeId(), ImmutableSet.of(partitionedSplit), true));
             }
 
-            // add unpartitioned sources
-            Optional<PlanNodeId> sourceId = driver.getSourceId();
-            if (sourceId.isPresent()) {
-                SplitAssignment splitAssignment = unpartitionedSplitAssignments.get(sourceId.get());
-                if (splitAssignment != null) {
-                    driver.updateSplitAssignment(splitAssignment);
-                }
-            }
-
             pendingCreations.decrementAndGet();
             closeDriverFactoryIfFullyCreated();
 
+            if (driverFactory.getSourceId().isPresent() && partitionedSplit == null) {
+                driverReferences.add(new WeakReference<>(driver));
+                scheduleSplits();
+            }
+
             return driver;
         }
 
+        public void enqueueSplits(Set<ScheduledSplit> splits, boolean noMoreSplits)
+        {
+            verify(driverFactory.getSourceId().isPresent(), "not a source driver");
+            queuedSplits.addAll(splits);
+            verify(!this.noMoreSplits.get() || noMoreSplits, "cannot unset noMoreSplits");
```

Comment:
Based on [this](https://github.com/trinodb/trino/blob/master/core/trino-main/src/main/java/io/trino/execution/SqlTaskExecution.java#L282) logic I assume the splits are delivered sequentially.

The logic is as following:

1. Every split has a sequence number assigned by the coordinator
2. Worker ignores splits with a sequence number lower than the acknowledged sequence number
3. Worker sets acknowledged sequence number to the highest sequence number in the last split batch

In my understanding if the splits were delivered out of order that would break the logic. For example if a batch of splits with the highest sequence number of `100` was delivered before a batch of splits with the highest sequence number of `50`, the later batch would get lost.

Reply:
Ah so the out of order problem is handled higher in the stack.  That makes sence then